python/semantic_kernel/connectors/memory/postgres/postgres_settings.py

```diff
@@ -19,4 +29,53 @@ class PostgresSettings(KernelBaseSettings):
 
     env_prefix: ClassVar[str] = "POSTGRES_"
 
-    connection_string: SecretStr
+    connection_string: SecretStr | None = None
+    host: str | None = Field(None, alias=PGHOST_ENV_VAR)
+    port: int | None = Field(5432, alias=PGPORT_ENV_VAR)
+    dbname: str | None = Field(None, alias=PGDATABASE_ENV_VAR)
+    user: str | None = Field(None, alias=PGUSER_ENV_VAR)
+    password: SecretStr | None = Field(None, alias=PGPASSWORD_ENV_VAR)
+    sslmode: str | None = Field(None, alias=PGSSL_MODE_ENV_VAR)
+
+    min_pool: int = 1
+    max_pool: int = 5
+
+    default_dimensionality: int = 100
+
+    def get_connection_args(self) -> dict[str, str | int | None]:
+        """Get connection arguments."""
+        result = conninfo_to_dict(self.connection_string.get_secret_value()) if self.connection_string else {}
+
+        if self.host:
+            result["host"] = self.host
+        if self.port:
+            result["port"] = self.port
+        if self.dbname:
+            result["dbname"] = self.dbname
+        if self.user:
+            result["user"] = self.user
+        if self.password:
+            result["password"] = self.password.get_secret_value()
+
+        # Ensure required values
+        if "host" not in result:
+            raise ValueError("host is required. Please set PGHOST or connection_string.")
+        if "dbname" not in result:
+            raise ValueError("database is required. Please set PGDATABASE or connection_string.")
+        if "user" not in result:
+            raise ValueError("user is required. Please set PGUSER or connection_string.")
+        if "password" not in result:
+            raise ValueError("password is required. Please set PGPASSWORD or connection_string.")
+
+        return result
+
+    async def create_connection_pool(self) -> AsyncConnectionPool:
```

Comment:
nit: I am not sure if the settings class should contain this method? Maybe move it to the utils.py?

Reply:
I do like these types of methods on settings, though I can see an argument against it. My thinking is when a user wants to transform settings into some instance of a type, it's more convenient to find and use when they don't have to import utils or discover it in documentation; it would be available through intellisense etc. It can sometimes cause import issues, but in this case it's just returning a psycopg connection pool which would otherwise be imported via the module initialization anyway.

In this case, creating a connection pool from settings is more important for users as the PostgresStore doesn't create it's own connection pool. This is because there's not a good way to manage the connection pool via the VectorStore abstraction. It's not a context manager, so it can't be explicitly disposed of. In other cases like the Qdrant store, the client is wrapped with a utiliy class that defines a `__del__` method that closes the connection on deletion of the object. I found that this technique did not work with the psycopg async connection pool - it would produce a lot of warnings about tasks in the asyncio queue not being completed before the loop was closed. So to avoid this, I force the user to pass in their own connection pool to the Vector Store. The VectorStoreRecordCollection doesn't have this issue, as it implements `__aenter__` and `__aexit__` and is intended to be used as a context manager.

That said, I'm happy to move the functionality to a utils method to fit better fit the project patterns. Let me know if so and I'll make the change.

Reply:
I see. I wasn't sure initially. I am good now that you have explained it and clearly you have thought about it.